processor/src/main/java/org/mapstruct/ap/internal/model/LifecycleCallbackFactory.java

```diff
@@ -125,8 +141,9 @@ private LifecycleCallbackFactory() {
                 ParameterAssignmentUtil.getParameterAssignments( availableParams, callback.getParameters() );
 
             if ( parameterAssignments != null
-                && callback.matches( extractSourceTypes( parameterAssignments ), method.getResultType() ) ) {
-
+                && callback.matches( extractSourceTypes( parameterAssignments ), method.getResultType() )
+                && ( callback.getReturnType().isVoid() || callback.getReturnType().isTypeVar()
+                    || callback.getReturnType().isAssignableTo( method.getResultType() ) ) ) {
```

Comment:
You'll also have to check if `method` has a non-void `returnType` (`!method.getReturnType().isVoid()`). Plus, I think in the assignable-to check, should also take `method.getReturnType()` into account instead of the `resultType` (the `resultType` may be more specific than `returnType` for certain mapping methods with `@MappingTarget`, but in most cases both will be the same).